major RISC-V assembly cleanups

npitre opened the pull request:
The risc-v assembly code is a mess of code duplication and inefficiencies.
Many things can be improved here.

This series removes a sizeable number of lines while improving the
existing functionality. And it is faster too, significantly so in some cases.

npitre committed:
riscv: use simplest asm expression when possible

Let's take advantage of assembler pseudoinstructions:

- convert `addi rd, rs, 0` to `mv rd, rs`
- convert `jal x0, somewhere` to `j somewhere`
- convert `csrrs x0, csrreg, rs` to `csrs csrreg, rs`
- convert `fscsr x0, rs` to `fscsr rs`

And simplify zero offsets to simply 0.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: fix wrong access width in assembly code

The thread->base.user_options field is an uint8_t. Access it using lb.
A "copy" of it is made into __esf.fp_state. Make that field an uint8_t
too and access it with lb/sb.

_callee_saved.fcsr is an uint32_t. Access it with lw/sw.
Ditto for is_user_mode.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: better abstraction for register-wide load/store opcodes

Those are prominent enough that having RV_OP_LOADREG and RV_OP_STOREREG
shouting at you all over the place is rather unpleasant and bad taste.

Let's create pseudo-instructions of our own with assembler macros
rather than preprocessor defines and only in assembly scope.
This makes the asm code way more uniform and readable.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: better abstraction for register-wide FP load/store opcodes

Same rationale as preceding commit. Let's create pseudo-instructions in
assembly scope to make the code more uniform and readable.

Furthermore the definition of COPY_ESF_FP() was wrong as the width of
floating point registers vary not according to CONFIG_64BIT but
CONFIG_CPU_HAS_FPU_DOUBLE_PRECISION. It is therefore wrong to use
lr/sr (previously RV_OP_LOADREG/RV_OP_STOREREG) and a regular temporary
register to transfer such content.

Note: There are far more efficient ways to copy FP context around but
      such optimisations will come separately.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: exception code mega simplification and optimization

Complete revamp of the exception entry code, including syscall handling.
Proper syscall frame exception trigger. Many correctness fixes, hacks
removal, etc. etc.

I tried to make this into several commits, but this stuff is all
inter-related and a pain to split.

The diffstat summary:

 14 files changed, 250 insertions(+), 802 deletions(-)

Binary size (before):

   text	   data	    bss	    dec	    hex	filename
   1104	      0	      0	   1104	    450	isr.S.obj
     64	      0	      0	     64	     40	userspace.S.obj

Binary size (after):

   text	   data	    bss	    dec	    hex	filename
    600	      0	      0	    600	    258	isr.S.obj
     36	      0	      0	     36	     24	userspace.S.obj

Run of samples/userspace/syscall_perf (before):

*** Booting Zephyr OS build zephyr-v3.0.0-325-g3748accae018  ***
Main Thread started; qemu_riscv32
Supervisor thread started
User thread started
Supervisor thread(0x80010048):       384 cycles	     509 instructions
User thread(0x80010140):           77312 cycles	   77437 instructions

Run of samples/userspace/syscall_perf (after):

*** Booting Zephyr OS build zephyr-v3.0.0-326-g4c877a2753b3  ***
Main Thread started; qemu_riscv32
Supervisor thread started
User thread started
Supervisor thread(0x80010048):       384 cycles	     509 instructions
User thread(0x80010138):            7040 cycles     7165 instructions

Yes, that's more than a 10x speed-up!

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
I added my big cleanup along with the other commits.

The major highlight is:

Run of samples/userspace/syscall_perf (before):

```
*** Booting Zephyr OS build zephyr-v3.0.0-325-g3748accae018  ***
Main Thread started; qemu_riscv32
Supervisor thread started
User thread started
Supervisor thread(0x80010048):       384 cycles      509 instructions
User thread(0x80010140):           77312 cycles    77437 instructions
```

Run of samples/userspace/syscall_perf (after):

```
*** Booting Zephyr OS build zephyr-v3.0.0-326-g4c877a2753b3  ***
Main Thread started; qemu_riscv32
Supervisor thread started
User thread started
Supervisor thread(0x80010048):       384 cycles      509 instructions
User thread(0x80010138):            7040 cycles     7165 instructions
```

Yes, that's more than a 10x speed-up!

This clashes big time with the recently merged SMP support. I do think it would
be far easier to port the SMP support on top of this rather than doing it the
other way around.

edersondisouza commented:
The first four patches - the simple refactor, look great. I didn't start looking at the "mega simplification" one yet. As it's fairly big, I'll probably do it on parts next week.

@npitre you mention about the conflicts with the SMP support series. Are you suggesting that, for your patches to go, the SMP series should be reverted and recreated in the future? I do understand that rebasing your series on top of current main is fairly complex now, but maybe the right tactics can help. I haven't looked carefully at the main patch yet - but I'll look at them thinking on ways to approaching this rebase. Hopefully there's a good way forward.

npitre commented:
> Are you suggesting that, for your patches to go, the SMP series should
> be reverted and recreated in the future? I do understand that rebasing
> your series on top of current main is fairly complex now, but maybe
> the right tactics can help. I haven't looked carefully at the main
> patch yet - but I'll look at them thinking on ways to approaching this
> rebase. Hopefully there's a good way forward.

I'm working on backporting SMP support on top of this at the moment.

npitre committed:
riscv: irq_offload: simpler implementation

Get rid of all those global variables and IRQ locking.
Use the regular IRQ exit path to let tests validate preemption properly.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: implement arch_switch()

The move to arch_switch() is a prerequisite for SMP support.

Make it optimal without the need for an ECALL roundtrip on every
context switch. Performance numbers from tests/benchmarks/sched:

Before:
unpend  107 ready  102 switch  188 pend  218 tot  615 (avg  615)

After:
unpend  107 ready  102 switch  170 pend  217 tot  596 (avg  595)

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: isr.S: optimize FP regs save/restore decision

Rely on mstatus rather than thread->base.user_options since it is always
up to date (updated by z_riscv_switch) to simplify the code and be SMP
proof. Also carry over SF_INIT to the mstatus being restored in case
it was changed in the mean time.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: align struct __esf properly

The minimum stack alignment is 16. Therefore, the stack space to store
a struct __esf object must be rounded up to the next 16-byte boundary.

It is not sufficient to do the rounding on the __z_arch_esf_t_SIZEOF
definition. When the stack is constructed in arch_new_thread() it is
also necessary to do the rounding there too.

Let's make the structure itself carry the alignment attribute instead to
make it work in all cases.

While at it, remove the unused _K_THREAD_NO_FLOAT_SIZEOF definition.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: move the tp register from caller-saved to callee-saved

This is a per-thread register that gets updated only when context
switching. No need to load and save it on every exception entry.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: isr.S: dedicate a register to &current_cpu

Stop using &_kernel as this is not SMP friendly. Let's use s0 (after
preserving its content) to hold &current_cpu instead so it won't have
to be reloaded 	each time it is needed. This will be even more relevant
when SMP support is added.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
Revert "arch/riscv: Do not use irq_lock() on arch_irq_offload"

This reverts commit b0458201ccd272f4d7e5b7fddbca952f77f67a3c.

The purpose of this commit will be reintroduced later on top of
a cleaner codebase.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
Revert "arch/riscv: Use arch_switch() for context swap"

This reverts commit be28de692c5e2c5546841c49d6a85d5f046474a4.

The purpose of this commit will be reintroduced later on top of
a cleaner codebase.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
Revert "arch/riscv: Get current CPU properly instead of assuming single CPU"

This reverts commit 8686ab54727c3d99d41a5de663656522557cf241.

The purpose of this commit will be reintroduced later on top of
a cleaner codebase.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
riscv: isr.S: compute _current_cpu using CPU number on SMP

To do so efficiently on systems without the mul instruction, we use
shifts and adds which is faster and sometimes smaller than a plain loop.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
SMP support added to this branch. I had to revert 3 existing commits
as they heavily conflicted with these cleanups and needed a different
implementation.

It is now equivalent in functionality (and in bugs) to the current
main branch. Yes, SMP in main has bugs whose fix will be nicer on top
of this.

npitre committed:
riscv: make arch_is_user_context() SMP compatible

This is painful. There is no way for u-mode code to know if we're
currently executing in u-mode without generating a fault, besides
stealing a general purpose register away from the standard ABI
that is. And a global variable doesn't work on SMP as this must be
per-CPU and we could be migrated to another CPU just at the right
moment to peek at the wrong CPU variable (and u-mode can't disable
preemption either).

So, given that we'll have to pay the price of an exception entry
anyway, let's at least make it free to privileged threads by using
the mscratch register as the non-user context indicator (it must
be zero in m-mode for exception entry to work properly). In the
case of u-mode we'll simulate a proper return value in the
exception trap code. Let's settle on the return value in t0
and omit the volatile to give the compiler a chance to cache
the result.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
OK, here it is done. And CI should be happy too.

The last 2 commits address issues that were not covered by the current
main branch's code. Most importantly is making `arch_is_user_context()`
SMP compatible.

nashif commented:
```qemu_riscv64_smp          tests/subsys/logging/log_stack/logging.log_stack_immediate_cpp FAILED Failed ```

Those tests just check stack sizes and it seems the new code change what the test is expecting. TBH, those tests are starting to be an annoyance, especially when working on bringing up a new architecture. @nordic-krch FYI

nordic-krch commented:
>  those tests are starting to be an annoyance

yes, noticed it couple of times already. On the other hand, logging is used everywhere so it's good to be track stack usage and delta. Things that could be tweaked there are:
- increase tolerance. Currently it is set to 16 bytes. Downside is that it will delay detection of increase.
- be more tolerant for NO_OPTIMIZATION case. I've seen that it tends to grow faster.

Other proposals?

fkokosinski commented on arch/riscv/core/irq_offload.c:
```diff
@@ -1,31 +1,13 @@
 /*
- * Copyright (c) 2016 Jean-Paul Etienne <fractalclone@gmail.com>
+ * Copyright (c) 2022 BayLibre SAS
```
Why is the reason behind this change in the copyrights?

fkokosinski commented on arch/riscv/core/asm_macros.inc:
```diff
@@ -32,3 +32,25 @@
 
 #endif
+
+#ifdef CONFIG_CPU_HAS_FPU_DOUBLE_PRECISION
```
If  another extension-related ifdef is being added, maybe it would be a good idea to name it as described [here](https://github.com/zephyrproject-rtos/zephyr/pull/43557/files#r822678126), i.e. `CONFIG_RISCV_ISA_EXT_D`?

fkokosinski replied:
Also, many empty lines could and probably should be removed from this file.

fkokosinski reviewed (commented):
I'm still looking at the changes introduced in f7b7be6037f439a435468da327f89b94da3ca50a, so for now I only have minor comments, but the rest looks promising and seems to have greatly improved the overall readability.

npitre commented:
> - * Copyright (c) 2016 Jean-Paul Etienne <fractalclone@gmail.com>
> + * Copyright (c) 2022 BayLibre SAS
> 
> Why is the reason behind this change in the copyrights?

The content of thisfile ended up being a complete rewrite.

> +#ifdef CONFIG_CPU_HAS_FPU_DOUBLE_PRECISION
> 
> If  another extension-related ifdef is being added, maybe it would 
> be a good idea to name it as described
> [here](https://github.com/zephyrproject-rtos/zephyr/pull/43557/files#r822678126),
> i.e. `CONFIG_RISCV_ISA_EXT_D`?

Indeed. Here I wanted to reuse what was there already to ease reviewing.
A separate commit on top of this to use better nomenclature would be good.

> Also, many empty lines could and probably should be removed from this file.

Style preference I guess. I don't have a strong opinion here though.

npitre committed:
riscv: implement CONFIG_IRQ_OFFLOAD_NESTED

It can easily be done now, so why not. Suffice to increment the nested
count like with actual IRQs.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
Latest changes:

- fixed exit path for irq_offload to consider preemption
- implemented CONFIG_IRQ_OFFLOAD_NESTED
- fixed tests/subsys/logging/log_stack after latest upstream changes

Can we please merge this now? SMP support in the main branch is buggy
and lacking (see the last 2 commits).

edersondisouza commented on arch/riscv/core/isr.S:
```diff
@@ -467,102 +423,23 @@ irq_done:
+	/* Perform context switch */
+	call z_riscv_switch
 
+z_riscv_thread_start:
```
Why this empty label?

edersondisouza reviewed (commented):
Looking ok, but didn't look into last patch - getting sleepy now.

npitre commented on include/arch/riscv/exp.h:
```diff
@@ -73,8 +73,11 @@ struct __esf {
 	ulong_t mstatus;	/* machine status register */
 
+#ifdef CONFIG_USERSPACE
+	ulong_t sp;		/* preserved (user or kernel) stack pointer */
```
Because when user space is not configured there are no different stacks to return to as the current stack pointer is always the one to use.

npitre commented on arch/riscv/core/isr.S:
```diff
@@ -60,102 +55,6 @@
-	lr temp, __z_arch_esf_t_a7_OFFSET(from_reg)		;\
-	sr temp, __z_arch_esf_t_a7_OFFSET(to_reg)
-
 #define DO_CALLEE_SAVED(op, reg) \
```
The action is actually the parameter in the form of the instruction to apply. I have no clear idea how this could be improved but if you think of something please tell me.

edersondisouza reviewed (commented):
Last patch quite deviates from the rest of the series, but I guess that's life. Looks good, but it would be nice if someone more acclimated with PMP helps review.

npitre commented:
On Tue, 15 Mar 2022, Ederson de Souza wrote:

> @edersondisouza commented on this pull request.

> Last patch quite deviates from the rest of the series, but I guess 
> that's life. Looks good, but it would be nice if someone more 
> acclimated with PMP helps review.

I deliberately didn't touch the PMP code, except for removing the PMP 
mapping of the `is_usermode` global variable when USERSPACE and SMP are 
enabled given such variable cannot be used in that case. So PMP remains 
unaffected by this series.

But PMP is on my list for a revamp as well.

The difficulty here is to find enough people with the skills set to 
review this code. The magnitude of this cleanup is a sign of that. So 
I'd advocate for a merge sooner than later given that remaining issues 
can be fixed later. This series does address many serious issues 
already so the greater risk is rather about not to merge it.

> I even wrote about the z_get_next_switch_handle issue, then noticed this is addressed at a later patch. Maybe squash this with previous one, so not to have known bad patches that may throw off some git bisect?

I prefer not for those reasons:

- It is still quite difficult to trigger the race.

- This is a workaround that deserves to be "documented" 
  separately to make it clearer to code archaeologists.

- SMP is obviously not fully functional until the very last patch so
  bisection is much more likely to fail for other reasons than this 
  anyway.

amergnat reviewed (commented):
Do you think it could be worth (for the next step/PR) using utvec register to have a dedicated userspace __irq_wrapper ?
I thing factorizing code used in both irq wrapper is possible but the code will maybe become less readable. The + is support a new privilege mode (i.e hypervisor) will be easier.

npitre commented on arch/riscv/core/isr.S:
That's not an address. The test is inlined all over the place. Here we test the opcode value to make sure we did hit the expected instruction. Please see the comment in arch_is_user_context()`.

amergnat replied:
Ack

npitre commented on arch/riscv/core/isr.S:
Agreed: the stack guard size should be big enough to accommodate the exception stack frame and some wiggle room. However this shouldn't be a merge criteria for this PR as CONFIG_PMP_STACK_GUARD=y doesn't build in main right now anyway.

amergnat replied:
Ack

MaureenHelm commented:
@npitre please rebase

@kgugala please review

npitre commented:
- Combined workaround for #40795 into the arch_switch commit as #43894
  to be merged soon clarifies things (this is lo longer a workaround).

- Rebased on latest main branch

fkokosinski reviewed (approved):
LGTM.

tovine commented on arch/riscv/core/asm_macros.inc:
```diff
@@ -0,0 +1,78 @@
+#ifdef CONFIG_64BIT
+	/* register-wide load/store based on ld/sd (XLEN = 64) */
+
+	.macro lr, rd, mem
```
Doesn't `lr` collide with the actual instruction named `lr` (load reserved, from the A extension)?
Technically it might be fine since the actual instruction mnemonics are `lr.w` and `lr.d` (for 32 and 64 bit values, respectively) but it can still lead to confusion...

npitre replied:
Hmmm... this could be confusing indeed. I wish risc-v had an official pseudo-op for that.
What about 'lx' and 'sx' instead since this is 'xlen' based? And  nqualified registers are identified with the letter x. Or any other suggestions?

tovine replied:
I think that sounds like a good idea. AFAIK neither `lx` nor `sx` is used for instruction names in any current or upcoming extensions.

npitre commented:
All tests pass now.
